tests/qasm/test_mpqp_to_qasm.py

```diff
@@ -0,0 +1,334 @@
+from mpqp.tools.circuit import random_circuit
+from mpqp.qasm.mpqp_to_qasm import mpqp_to_qasm2
+from mpqp.qasm.open_qasm_2_and_3 import remove_user_gates
+
+
+@pytest.mark.parametrize(
+    "instructions",
+    [
+        [X(0), Y(1), Z(2), H(3), CNOT(0, 1), CNOT(1, 2), BasisMeasure([0])],
+        [
+            X(0),
+            X(1),
+            X(2),
+            X(3),
+            Y(1),
+            Z(2),
+            H(3),
+            CNOT(0, 1),
+            CNOT(1, 2),
+            BasisMeasure(),
+        ],
+        [
+            S(0),
+            X(0),
+            Y(0),
+            Z(0),
+            P(np.pi, 0),
+            U(np.pi, np.pi / 2, 2.5, 0),
+            T(0),
+            CNOT(0, 1),
+            CRk(1, 1, 0),
+            Rk(1, 1),
+            CZ(0, 1),
+            Rx(0, 1),
+            Ry(0, 1),
+            Rz(0, 1),
+            TOF([0, 1], 2),
+            BasisMeasure([0, 1, 2]),
+        ],
+        [
+            P(np.pi, 0),
+            P(np.pi, 1),
+            P(np.pi / 2, 0),
+            P(np.pi, 1),
+            P(np.pi, 1),
+            P(np.pi, 0),
+            P(np.pi, 1),
+        ],
+        [
+            P(np.pi, 0),
+            P(np.pi, 1),
+            P(np.pi / 2, 0),
+            P(np.pi, 1),
+            P(np.pi, 1),
+            P(np.pi, 0),
+            P(np.pi, 2),
+        ],
+        [X(0), X(0), X(1), Y(1), Y(0), Y(1), Y(0), Z(3)],
+        [X(0), X(0), X(1), Barrier(), Y(1), Y(0), Y(0), Y(1)],
+        [
+            S(0),
+            X(0),
+            Y(0),
+            Z(0),
+            P(np.pi, 0),
+            U(np.pi, np.pi / 2, 2.5, 0),
+            T(0),
+            CNOT(0, 1),
+            CRk(1, 1, 0),
+            Rk(1, 1),
+            CZ(0, 1),
+            Rx(0, 1),
+            Ry(0, 1),
+            Rz(0, 1),
+            TOF([0, 1], 2),
+        ],
+        [Barrier()],
+        [X(0), Y(1), BasisMeasure([0])],
+        [CNOT(1, 0), CNOT(0, 1)],
+    ],
+)
+def test_mpqp_to_qasm(instructions: list[Instruction]):
+    circuit = QCircuit(instructions)
+    from qiskit import qasm2, QuantumCircuit
+
+    qiskit_circuit = circuit.to_other_language(Language.QISKIT)
+    assert isinstance(qiskit_circuit, QuantumCircuit)
+    assert qasm2.dumps(qiskit_circuit) == circuit.to_other_language(Language.QASM2)
+
+
+@pytest.mark.parametrize(
+    "instructions",
+    [
+        [
+            CustomGate(
+                UnitaryMatrix(
+                    np.array([[0, 1, 0, 0], [0, 0, 0, 1], [1, 0, 0, 0], [0, 0, 1, 0]])
+                ),
+                [1, 2],
+            )
+        ]
+    ],
+)
+def test_mpqp_to_qasm_custom_gate(instructions: list[Instruction]):
+    circuit = QCircuit(instructions)
+    from qiskit import qasm2, QuantumCircuit
+
+    qiskit_circuit = circuit.to_other_language(Language.QISKIT)
+    assert isinstance(qiskit_circuit, QuantumCircuit)
+    str_qiskit_circuit = remove_user_gates(qasm2.dumps(qiskit_circuit), True)
```

Comment:
this looks like circular logic to me, aren't we doing this exact same line in `o_other_language(Language.QASM2)` ?

Reply:
No, we use the mpqp_to_qasm2 method. But if you prefer, we can hard-code the test ?

Reply:
done